Fix for some non-linux

yangfl opened the pull request.

yangfl committed:
Fix fisdir sigfault on Hurd

Apparently Hurd's stat() does not like NULL and will sigfault, this patch takes
care of it.

yangfl committed:
test: use sleep() for mtime delay

For filesystems without high-resolution time support, an usleep() is too short
to alter mtime. Though 1s is quite safe, sleep for 2s in case it gets awaken
right within that second.

troglobit commented:
OK, so far this looks perfectly fine.  Happy to see a Hurd commit after such a long time away from it :) 

Just drop me a comment when you're done with the WIP (can't merge yet due to that).

Cheers
 /Joachim

yangfl committed:
Replace mknod() with creat()

For mknod(), mode must be S_IFIFO otherwise the behaviour is unspecified
according to SUS, and apparently Debian/kFreeBSD is in the hater list.

troglobit commented on src/lite.h:
```diff
@@ -129,8 +129,9 @@ char   *which      (const char *cmd);
 #ifndef touch
-#include <sys/stat.h>		/* utimensat() */
-#include <sys/time.h>		/* utimensat() on *BSD */
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <utime.h>
 /**
```
I need to check this with all the supported targets, but otherwise tentatively approve.

troglobit commented on src/lite.h:
```diff
@@ -139,10 +140,15 @@ int     whichp     (const char *cmd);
-		if (errno == ENOENT)
-			return mknod((path), S_IFREG|0644, 0);
-		return -1;
+	int fd;
+
+	if (utime(path, NULL)) {
+		if (errno != ENOENT)
+			return -1;
+		fd = creat(path, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
+		if (fd < 0)
+			return -1;
+		close(fd);
```
This I cannot approve, unfortunately.  We cannot rely on the poor granularity of `utime()` on the platforms that use `touch()` in Finit. Sorry, but we've had some really nasty bugs with this in the past.  Please replace this with [inadyn/lib/utimensat.c ](https://github.com/troglobit/inadyn/blob/master/lib/utimensat.c), which I should've added myself years ago to libite, but never got around to.

Not really sure what's the most appropriate way, but since it's in POSIX.1-2008 I think as a library replacement is the best way.  I.e., exactly like it's done in In-a-dyn (and mdnsd, ssdp-responder, bridged, mini-snmpd, mrouted, smcroute) and other's I've run into this problem with in the past.

troglobit commented on src/pidfile.c:
```diff
@@ -38,14 +38,13 @@
 
-#include <sys/stat.h>		/* utimensat() */
-#include <sys/time.h>		/* utimensat() on *BSD */
 #include <sys/types.h>
 #include <errno.h>
 #include <paths.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <unistd.h>
+#include <utime.h>
```
I need to check this with all the supported targets, but otherwise tentatively approve.

troglobit commented on src/pidfile.c:
```diff
@@ -83,7 +82,7 @@ int pidfile(const char *basename)
 		if (!access(pidfile_path, R_OK) && pid == pidfile_pid) {
-			utimensat(0, pidfile_path, NULL, 0);
+			utime(pidfile_path, NULL);
 			return (0);
```
Same as above, must use `utimensat()`.

troglobit commented on test/pidfile.c:
```diff
@@ -41,7 +41,7 @@ static int mtime()
 
 	/* Must sleep a while here otherwise we execute too fast => no mtime change :-( */
-	usleep(10000);
+	sleep(2);
 
```
Test, this is fine :)

troglobit commented on test/touch.c:
```diff
@@ -23,7 +23,7 @@ static int mtime(char *file)
 
 	/* Must sleep a while here otherwise we execute too fast => no mtime change :-( */
-	usleep(10000);
+	sleep(2);
```
Test, this is fine :)

Although, in the commit message, could you please elaborate why the number 2 was selected.

troglobit commented on src/fisdir.c:
```diff
@@ -41,7 +41,7 @@ int fisdir(const char *path)
 	struct stat sb;
 
-	if (!stat(path, &sb) && S_ISDIR(sb.st_mode))
+	if (path && !stat(path, &sb) && S_ISDIR(sb.st_mode))
```
Awesome catch, thanks!

troglobit commented:
Ah, now I see, you dropped the `utimensat()` removal and force pushed.  OK, I'll merge this, I assume you'll rework the other parts for FreeBSD somehow?

yangfl commented:
Since you mentioned inadyn/lib/utimensat.c I'll patch that downstream.
